File: solutions/nim/linked-list/1/linked_list.nim
# Dedicated to Shree DR.MDD
type
  Node[T] = ref object
    data: T
    back: Node[T]
    forth: Node[T]

  LinkedList*[T] = object
    head: Node[T]
    tail: Node[T]

proc len*[T](lst: LinkedList[T]): int =
  var iter = lst.head
  while iter != nil:
    inc result
    iter = iter.forth

proc push*[T](lst: var LinkedList[T], data: T) =
  var n = Node[T](data: data, back: lst.tail, forth: nil)
  if n.back != nil:
    n.back.forth = n
  if lst.head == nil:
    lst.head = n
  lst.tail = n

proc pop*[T](lst: var LinkedList[T]): T =
  let gone = lst.tail
  lst.tail = gone.back
  if lst.tail == nil:
    lst.head = nil
  else:
    lst.tail.forth = nil
  gone.data

proc shift*[T](lst: var LinkedList[T]): T =
  let gone = lst.head
  lst.head = gone.forth
  if lst.head != nil:
    lst.head.back = nil
  else:
    lst.tail = nil
  gone.data

proc unshift*[T](lst: var LinkedList[T], data: T) =
  var n = Node[T](data: data, back: nil, forth: lst.head)
  if n.forth != nil:
    n.forth.back = n
  lst.head = n
  if lst.tail == nil:
    lst.tail = n

proc delete*[T](lst: var LinkedList[T], data: T) =
  var iter = lst.head
  while iter != nil:
    if iter.data == data:
      if iter == lst.head:
        lst.head = iter.forth
      else:
        iter.back.forth = iter.forth
      if iter == lst.tail:
        lst.tail = iter.back
      else:
        iter.forth.back = iter.back
      return
    iter = iter.forth

File: solutions/javascript/linked-list/1/linked-list.js
//
// For my Shree DR.MDD
//
class ListNode {
  constructor(data) {
    this.data = data;
    this.next = null;
    this.prev = null;
  }
}

export class LinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  push(data) {
    const node = new ListNode(data);

    if (!this.tail) {
      this.head = node;
      this.tail = node;
    } else {
      this.tail.next = node;
      node.prev = this.tail;
      this.tail = node;
    }

    this.length++;
  }

  pop() {
    if (!this.tail) return null;

    const node = this.tail;
    if (this.tail.prev) {
      this.tail = this.tail.prev;
      this.tail.next = null;
    } else {
      this.head = null;
      this.tail = null;
    }

    this.length--;
    return node.data;
  }

  shift() {
    if (!this.head) return null;

    const node = this.head;
    if (this.head.next) {
      this.head = this.head.next;
      this.head.prev = null;
    } else {
      this.head = null;
      this.tail = null;
    }

    this.length--;
    return node.data;
  }

  unshift(data) {
    const node = new ListNode(data);

    if (!this.head) {
      this.head = node;
      this.tail = node;
    } else {
      node.next = this.head;
      this.head.prev = node;
      this.head = node;
    }
    this.length++;
  }

  delete(data) {
    if (!this.head) return;
    let current = this.head;

    while (current) {
      if (current.data === data) {
        if (current.prev) {
          current.prev.next = current.next;
        } else {
          this.head = current.next;
        }

        if (current.next) {
          current.next.prev = current.prev;
        } else {
          this.tail = current.prev;
        }

        this.length--;
        return;
      }
      current = current.next;
    }
  }

  count() {
    return this.length;
  }
}
